In [1]:
import os

# --- CONFIGURATION (REMPLIS ÇA) ---
GIT_USER = "moubarik-ashraf"
GIT_EMAIL = "moubarikashraf@gmail.com"
TOKEN = "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" # Ton Personal Access Token
REPO_NAME = "DS_Project" # Le nom exact du repo créé à l'étape 1


# --- CLONAGE DU PROJET ---
if not os.path.exists(REPO_NAME):
    print("📥 Récupération du projet depuis GitHub...")
    # On clone avec le token pour avoir le droit de modifier (push) plus tard
    !git clone https://{TOKEN}@github.com/{GIT_USER}/{REPO_NAME}.git
else:
    print("✅ Le projet est déjà là.")

# --- SE PLACER DANS LE PROJET ---
os.chdir(REPO_NAME)
print(f"📍 Dossier de travail actuel : {os.getcwd()}")

# Vérification que tout est là
print("\n📂 Contenu du dossier :")
!ls -R

✅ Le projet est déjà là.
📍 Dossier de travail actuel : /content/DS_Project

📂 Contenu du dossier :
.:
data  models  notebooks  reports  sample_data  src

./data:
processed  raw

./data/processed:

./data/raw:
DATA_students.csv  description.csv

./models:

./notebooks:
01_exploration.ipynb  02_preprocessing.ipynb  03_modeling.ipynb

./reports:

./sample_data:

./src:
utils.py


In [2]:
import pandas as pd

# Chemin vers le fichier
file_path = 'data/raw/DATA_students.csv'

# Lire le fichier CSV
df = pd.read_csv(file_path)

# Afficher les 5 premières lignes
df.head()

payment_type  profit_per_order  sales_per_customer  category_id  \
0      PAYMENT               NaN           195.02570         46.0   
1     TRANSFER         85.423610           245.20793          NaN   
2      PAYMENT        261.173770           456.55527          9.0   
3        DEBIT        -52.374670           191.65901         48.0   
4        DEBIT         55.085342           187.45561         48.0   

          category_name customer_city customer_country  customer_id  \
0  Indoor/Outdoor Games       Milford          EE. UU.    3736.8790   
1  Indoor/Outdoor Games        Caguas      Puerto Rico    4268.8580   
2      Cardio Equipment        Caguas      Puerto Rico    7772.3945   
3          Water Sports        Caguas      Puerto Rico    9977.6810   
4                   NaN        Caguas      Puerto Rico    8920.0180   

  customer_segment customer_state  ...    order_region    order_state  \
0              NaN             CT  ...  Western Europe         Vienna   
1         Consumer             PR  ...   South America   Buenos Aires   
2         Consumer             PR  ...  Western Europe  West Flanders   
3         Consumer             PR  ...   South America      O'Higgins   
4         Consumer             PR  ...     East of USA       New York   

      order_status  product_card_id  product_category_id  \
0              NaN             1014                 46.0   
1          PENDING             1014                 46.0   
2  PENDING_PAYMENT              191                  9.0   
3         COMPLETE             1073                 48.0   
4          ON_HOLD             1073                 48.0   

                        product_name product_price shipping_date  \
0   O'Brien Men's Neoprene Life Vest         49.98     42322.746   
1   O'Brien Men's Neoprene Life Vest         49.98     42340.965   
2  Nike Men's Free 5.0+ Running Shoe         99.99     42061.550   
3        Pelican Sunstream 100 Kayak        199.99     42883.145   
4        Pelican Sunstream 100 Kayak        199.99     42950.610   

    shipping_mode  delivery_outcome  
0  Standard Class           on time  
1  Standard Class           delayed  
2  Standard Class     early arrival  
3  Standard Class     early arrival  
4  Standard Class           on time  

[5 rows x 41 columns]

In [3]:
print(f"Le DataFrame compte {df.shape[0]} lignes et {df.shape[1]} colonnes.")

Le DataFrame compte 15548 lignes et 41 colonnes.


In [4]:
print("Répartition de la colonne 'delivery_outcome' :\n")

delivery_outcome_counts = df['delivery_outcome'].value_counts()
delivery_outcome_percentages = df['delivery_outcome'].value_counts(normalize=True) * 100

delivery_outcome_distribution = pd.DataFrame({
    'Effectifs': delivery_outcome_counts,
    'Pourcentage': delivery_outcome_percentages
})

display(delivery_outcome_distribution)

Répartition de la colonne 'delivery_outcome' :



Effectifs  Pourcentage
delivery_outcome                        
delayed                8909    57.299974
early arrival          3714    23.887317
on time                2925    18.812709

In [5]:
import plotly.express as px

# Créer un diagramme circulaire interactif avec Plotly
fig = px.pie(
    delivery_outcome_distribution,
    names=delivery_outcome_distribution.index, # Les catégories sont l'index du DataFrame
    values='Effectifs',
    title='Distribution du Résultat de Livraison (Delivery Outcome)',
    hole=0.3, # Pour un diagramme en anneau (donut chart), optionnel
    labels={'names': 'Résultat de Livraison', 'Effectifs': 'Nombre de Commandes'}
)

# Mettre à jour le texte affiché sur les tranches pour inclure les pourcentages
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.update_layout(uniformtext_minsize=12, uniformtext_mode='hide')

fig.show()

In [6]:
print("Types de données par colonne :")
print(df.dtypes)

print("\nRésumé des colonnes numériques :")
print(df.select_dtypes(include=['number']).columns.tolist())

print("\nRésumé des colonnes catégorielles (object) :")
print(df.select_dtypes(include=['object']).columns.tolist())

# Pour les dates, on peut essayer d'identifier des colonnes basées sur le nom ou convertir et voir les erreurs
# Pour une détection plus robuste des dates, il faudrait une analyse plus approfondie ou une conversion explicite
date_cols = [col for col in df.columns if 'date' in col.lower()]
if date_cols:
    print("\nColonnes potentiellement de type date (basé sur le nom) :")
    print(date_cols)
else:
    print("\nAucune colonne potentiellement de type date trouvée (basé sur le nom).")

Types de données par colonne :
payment_type                 object
profit_per_order            float64
sales_per_customer          float64
category_id                 float64
category_name                object
customer_city                object
customer_country             object
customer_id                 float64
customer_segment             object
customer_state               object
customer_zipcode            float64
department_id               float64
department_name              object
latitude                    float64
longitude                   float64
market                       object
order_city                   object
order_country                object
order_customer_id           float64
order_date                  float64
order_id                    float64
order_item_cardprod_id      float64
order_item_discount         float64
order_item_discount_rate    float64
order_item_id               float64
order_item_product_price    float64
order_item_profit_ratio     float

In [7]:
missing_data = df.isnull().sum()
missing_percentage = round((df.isnull().sum() / len(df)) * 100,2)
unique_values_count = df.nunique()

missing_info = pd.DataFrame({
    'Valeurs manquantes': missing_data,
    '% de manquants': missing_percentage,
    'Nombre de valeurs uniques': unique_values_count
})

# Suppression du filtrage pour afficher toutes les colonnes, y compris celles avec 0% de manquants
missing_info = missing_info.sort_values(
    by='% de manquants', ascending=False
)

print("Variables avec des valeurs manquantes, leur taux, et le nombre de valeurs uniques (0% si aucun manquant) :\n")
display(missing_info)

Variables avec des valeurs manquantes, leur taux, et le nombre de valeurs uniques (0% si aucun manquant) :



Valeurs manquantes  % de manquants  \
order_item_total_amount                  854            5.49   
department_id                            820            5.27   
order_profit_per_order                   804            5.17   
order_region                             799            5.14   
customer_id                              798            5.13   
market                                   795            5.11   
customer_city                            793            5.10   
category_id                              791            5.09   
shipping_mode                            787            5.06   
order_item_discount                      786            5.06   
payment_type                             779            5.01   
order_status                             777            5.00   
order_item_discount_rate                 776            4.99   
sales_per_customer                       776            4.99   
order_item_quantity                      774            4.98   
sales                                    771            4.96   
order_item_profit_ratio                  769            4.95   
order_state                              770            4.95   
latitude                                 767            4.93   
order_item_product_price                 763            4.91   
product_price                            764            4.91   
product_category_id                      764            4.91   
category_name                            762            4.90   
customer_country                         757            4.87   
customer_zipcode                         750            4.82   
order_country                            744            4.79   
profit_per_order                         742            4.77   
longitude                                742            4.77   
customer_state                           737            4.74   
customer_segment                         736            4.73   
order_city                               736            4.73   
product_name                             733            4.71   
department_name                          725            4.66   
order_item_id                              0            0.00   
order_item_cardprod_id                     0            0.00   
order_date                                 0            0.00   
order_id                                   0            0.00   
order_customer_id                          0            0.00   
product_card_id                            0            0.00   
shipping_date                              0            0.00   
delivery_outcome                           0            0.00   

                          Nombre de valeurs uniques  
order_item_total_amount                        1510  
department_id                                    11  
order_profit_per_order                         9679  
order_region                                     23  
customer_id                                   14682  
market                                            5  
customer_city                                   558  
category_id                                      51  
shipping_mode                                     4  
order_item_discount                            1177  
payment_type                                      4  
order_status                                      7  
order_item_discount_rate                        235  
sales_per_customer                             8510  
order_item_quantity                               6  
sales                                           453  
order_item_profit_ratio                        1863  
order_state                                     890  
latitude                                      14263  
order_item_product_price                        374  
product_price                                   375  
product_category_id                              51  
category_name                                    50  
customer_country                         

## check what is sales_per_customer

In [8]:
# 1. On nettoie d'abord : On ne garde que les lignes où 'sales_per_customer' existe
df_clean = df.dropna(subset=['sales_per_customer'])

# 2. On cherche parmi ces lignes propres un client qui a au moins 3 commandes
# (value_counts donne le nombre de commandes par client)
top_customers = df_clean['customer_id'].value_counts()
candidate_id = top_customers.index[0] # On prend le premier de la liste "propre"

print(f"🕵️‍♂️ Nouveau candidat analysé : ID {candidate_id}")
print(f"📦 Nombre de commandes analysables : {top_customers.iloc[0]}")

# 3. On récupère son historique
cols = ['order_date', 'order_item_total_amount', 'sales', 'sales_per_customer']
history = df[df['customer_id'] == candidate_id][cols].sort_values('order_date')

print("\n--- HISTORIQUE DÉTAILLÉ ---")
display(history)

# 4. LE TEST DE VÉRITÉ
# Hypothèse A : Est-ce la somme historique exacte ?
total_paid_real = history['order_item_total_amount'].sum()
val_displayed = history['sales_per_customer'].iloc[0] # On regarde la valeur affichée

print("\n--- VERDICT ---")
print(f"💰 Total réellement payé (Somme de la colonne 'Total Amount') : {total_paid_real:.2f}")
print(f"❓ Valeur affichée dans 'sales_per_customer' : {val_displayed:.2f}")

# Hypothèse B : Est-ce une constante ou une valeur qui change ?
if history['sales_per_customer'].nunique() == 1:
    print(" La valeur est CONSTANTE sur toutes les lignes.")
else:
    print("La valeur EVOLUE à chaque commande.")

# Hypothèse C : Est-ce juste une copie de la vente actuelle (Doublon) ?
# On vérifie si la colonne A est égale à la colonne B ligne par ligne
is_duplicate = (history['sales_per_customer'] == history['sales']).all() or \
               (history['sales_per_customer'] == history['order_item_total_amount']).all()

if is_duplicate:
    print("C'est juste une copie de la vente actuelle (Doublon inutile).")

🕵️‍♂️ Nouveau candidat analysé : ID 2553.9749
📦 Nombre de commandes analysables : 2

--- HISTORIQUE DÉTAILLÉ ---


order_date  order_item_total_amount   sales  sales_per_customer
9671    42856.234                 122.9508  149.94              122.84
12167   42888.990                 175.9912  199.99              179.99


--- VERDICT ---
💰 Total réellement payé (Somme de la colonne 'Total Amount') : 298.94
❓ Valeur affichée dans 'sales_per_customer' : 122.84
La valeur EVOLUE à chaque commande.


## Product category name/id imputing

In [9]:
print("Avant l'imputation des catégories des produits :")
print(df[['category_id', 'category_name']].isnull().sum())

# 1. Créer un mappage de ID vers Nom et de Nom vers ID
id_to_name_cat = df.dropna(subset=['category_id', 'category_name'])\
                 .set_index('category_id')['category_name'].drop_duplicates().to_dict()
name_to_id_cat = df.dropna(subset=['category_id', 'category_name'])\
                 .set_index('category_name')['category_id'].drop_duplicates().to_dict()

# 2. Imputer les category_id manquants si category_name est présent
def fill_category_id(row):
    if pd.isnull(row['category_id']) and pd.notnull(row['category_name']):
        return name_to_id_cat.get(row['category_name'], row['category_id'])
    return row['category_id']

df['category_id'] = df.apply(fill_category_id, axis=1)

# 3. Imputer les category_name manquants si category_id est présent
def fill_category_name(row):
    if pd.isnull(row['category_name']) and pd.notnull(row['category_id']):
        return id_to_name_cat.get(row['category_id'], row['category_name'])
    return row['category_name']

df['category_name'] = df.apply(fill_category_name, axis=1)

print("\nAprès l'imputation des catégories :")
print(df[['category_id', 'category_name']].isnull().sum())

# Vérifier si toutes les valeurs manquantes de l'une ou l'autre colonne ont été remplies
# Note: Il peut rester des NaNs si un ID/Nom était manquant et n'avait pas de correspondance dans le mapping.
missing_after_imputation_id_cat = df['category_id'].isnull().sum()
missing_after_imputation_name_cat = df['category_name'].isnull().sum()

if missing_after_imputation_id_cat == 0 and missing_after_imputation_name_cat == 0:
    print("\nImputation des colonnes 'category_id' et 'category_name' réussie : toutes les valeurs manquantes liées ont été remplies.")
elif missing_after_imputation_id_cat < 791 or missing_after_imputation_name_cat < 762:
    print(f"\nL'imputation a permis de réduire les valeurs manquantes dans 'category_id' de 791 à {missing_after_imputation_id_cat} et dans 'category_name' de 762 à {missing_after_imputation_name_cat}.")
else:
    print("\nL'imputation n'a pas pu remplir toutes les valeurs manquantes, car certaines n'avaient pas de correspondance dans le mapping initial.")

Avant l'imputation des catégories des produits :
category_id      791
category_name    762
dtype: int64

Après l'imputation des catégories :
category_id      45
category_name    53
dtype: int64

L'imputation a permis de réduire les valeurs manquantes dans 'category_id' de 791 à 45 et dans 'category_name' de 762 à 53.


## Departement name/id imputing

In [10]:

# --- 1. ETAT INITIAL ---
cols_dept = ['department_id', 'department_name']
print("Manquants avant imputation :")
print(df[cols_dept].isnull().sum())

# --- 2. SYNCHRONISATION ID <-> NOM (Interne) ---
# Si on a le Nom mais pas l'ID (ou l'inverse), on bouche le trou.

# Création des dictionnaires de référence
# On nettoie les doublons pour avoir une relation 1 pour 1 propre
ref_df = df.dropna(subset=['department_id', 'department_name']).drop_duplicates(subset=['department_id'])
id_to_name = ref_df.set_index('department_id')['department_name'].to_dict()
name_to_id = ref_df.set_index('department_name')['department_id'].to_dict()

# Remplissage croisé (beaucoup plus rapide que .apply)
df['department_name'] = df['department_name'].fillna(df['department_id'].map(id_to_name))
df['department_id'] = df['department_id'].fillna(df['department_name'].map(name_to_id))

print(f"Après synchro ID/Nom : {df[cols_dept].isnull().sum().sum()} manquants restants")

# --- 3. IMPUTATION VIA LE PRODUIT (La méthode forte) ---
# Le 'product_card_id' détermine forcément le 'department_id'.

# On crée un mapping : Produit -> Département ID
prod_to_dept_id = df.dropna(subset=['product_card_id', 'department_id'])\
                    .drop_duplicates(subset=['product_card_id'])\
                    .set_index('product_card_id')['department_id']\
                    .to_dict()

# On crée un mapping : Produit -> Département Name
prod_to_dept_name = df.dropna(subset=['product_card_id', 'department_name'])\
                      .drop_duplicates(subset=['product_card_id'])\
                      .set_index('product_card_id')['department_name']\
                      .to_dict()

# Application de l'imputation
df['department_id'] = df['department_id'].fillna(df['product_card_id'].map(prod_to_dept_id))
df['department_name'] = df['department_name'].fillna(df['product_card_id'].map(prod_to_dept_name))
print(f"Après Imputation via le produit : {df[cols_dept].isnull().sum().sum()} manquants restants")

# --- 4. RE-SYNCHRONISATION ID <-> NOM (Interne) ---
# Si on a le Nom mais pas l'ID (ou l'inverse), on bouche le trou.

# Création des dictionnaires de référence
# On nettoie les doublons pour avoir une relation 1 pour 1 propre
ref_df = df.dropna(subset=['department_id', 'department_name']).drop_duplicates(subset=['department_id'])
id_to_name = ref_df.set_index('department_id')['department_name'].to_dict()
name_to_id = ref_df.set_index('department_name')['department_id'].to_dict()

# Remplissage croisé (beaucoup plus rapide que .apply)
df['department_name'] = df['department_name'].fillna(df['department_id'].map(id_to_name))
df['department_id'] = df['department_id'].fillna(df['department_name'].map(name_to_id))

print(f"Après 2ème synchro ID/Nom : {df[cols_dept].isnull().sum().sum()} manquants restants")


# --- 4. RAPPORT FINAL ---
print("\n--- RÉSULTAT FINAL ---")
print(df[cols_dept].isnull().sum())

# Vérification s'il reste des irréductibles
if df['department_id'].isnull().sum() > 0:
    print("\n⚠️ Il reste des départements vides. Cela signifie que certains 'product_card_id' sont orphelins (jamais associés à un département connu).")
else:
    print("\n✅ Tous les départements ont été retrouvés grâce aux produits !")

Manquants avant imputation :
department_id      820
department_name    725
dtype: int64
Après synchro ID/Nom : 66 manquants restants
Après Imputation via le produit : 0 manquants restants
Après 2ème synchro ID/Nom : 0 manquants restants

--- RÉSULTAT FINAL ---
department_id      0
department_name    0
dtype: int64

✅ Tous les départements ont été retrouvés grâce aux produits !


In [11]:


# --- 1. VERIFICATION INITIALE ---
miss_lat_start = df['latitude'].isnull().sum()
miss_lon_start = df['longitude'].isnull().sum()

print(f"🛑 Manquants DÉPART : latitude = {miss_lat_start} | longitude = {miss_lon_start}")

# --- FONCTION UTILITAIRE : MODE PAR GROUPE ---
def get_mode_coords_map(df, group_col):
    """
    Crée un dictionnaire {ID : (Lat, Lon)} basé sur la localisation la plus fréquente.
    """
    # On ne travaille que sur les lignes complètes pour construire la référence
    temp = df.dropna(subset=['latitude', 'longitude']).copy()
    temp['gps_tuple'] = list(zip(temp['latitude'], temp['longitude']))

    mode_map = temp.groupby(group_col)['gps_tuple']\
                   .agg(lambda x: x.mode()[0] if not x.mode().empty else None)\
                   .to_dict()
    return mode_map

# --- ÉTAPE 1 : IMPUTATION PAR PRODUIT (Priorité Absolue) ---
# On récupère les coordonnées "Standard" pour chaque produit
prod_gps_map = get_mode_coords_map(df, 'product_card_id')

lat_map_prod = {k: v[0] for k, v in prod_gps_map.items() if v}
lon_map_prod = {k: v[1] for k, v in prod_gps_map.items() if v}

# On remplit les trous indépendamment (si j'ai la longitude, je ne touche qu'à la latitude)
df['latitude'] = df['latitude'].fillna(df['product_card_id'].map(lat_map_prod))
df['longitude'] = df['longitude'].fillna(df['product_card_id'].map(lon_map_prod))

print(f"📉 Après Produit : latitude = {df['latitude'].isnull().sum()} | longitude = {df['longitude'].isnull().sum()}")

# --- ÉTAPE 2 : IMPUTATION PAR DÉPARTEMENT (Secours) ---
# Pour les produits inconnus, on regarde le rayon
dept_gps_map = get_mode_coords_map(df, 'department_id')

lat_map_dept = {k: v[0] for k, v in dept_gps_map.items() if v}
lon_map_dept = {k: v[1] for k, v in dept_gps_map.items() if v}

df['latitude'] = df['latitude'].fillna(df['department_id'].map(lat_map_dept))
df['longitude'] = df['longitude'].fillna(df['department_id'].map(lon_map_dept))

print(f"📉 Après Département : latitude = {df['latitude'].isnull().sum()} | longitude = {df['longitude'].isnull().sum()}")

# --- ÉTAPE 3 : FILET DE SÉCURITÉ (Fallback Global) ---
# S'il reste des trous dans l'un ou l'autre
if df['latitude'].isnull().sum() > 0:
    top_lat = df['latitude'].mode()[0]
    df['latitude'] = df['latitude'].fillna(top_lat)
    print("🧹 latitude : Remplissage final avec le Mode Global.")

if df['longitude'].isnull().sum() > 0:
    top_lon = df['longitude'].mode()[0]
    df['longitude'] = df['longitude'].fillna(top_lon)
    print("🧹 longitude : Remplissage final avec le Mode Global.")

# --- RAPPORT FINAL ---
print("-" * 30)
print(f"✅ RÉSULTAT FINAL :")
print(f"   latitude  Manquants : {df['latitude'].isnull().sum()}")
print(f"   longitude Manquants : {df['longitude'].isnull().sum()}")

🛑 Manquants DÉPART : latitude = 767 | longitude = 742
📉 Après Produit : latitude = 0 | longitude = 0
📉 Après Département : latitude = 0 | longitude = 0
------------------------------
✅ RÉSULTAT FINAL :
   latitude  Manquants : 0
   longitude Manquants : 0


# Task
Re-examine missing values and unique counts for 'customer_city', 'customer_country', 'customer_zipcode', and 'customer_state'; then, impute missing 'customer_city', 'customer_state', and 'customer_country' values using mappings from 'customer_zipcode', followed by imputing 'customer_country' from 'customer_city' and 'customer_state' combinations. Finally, review the remaining missing values and summarize the imputation results.

## Analyze Customer Location Missing Values

### Subtask:
Re-examine the missing values and unique counts for 'customer_city', 'customer_country', 'customer_zipcode', and 'customer_state' to understand the current state before imputation.


**Reasoning**:
To understand the current state of customer location data before further imputation, I will calculate and display the missing values and unique counts for the specified columns: 'customer_city', 'customer_country', 'customer_zipcode', and 'customer_state'.



In [12]:
customer_location_cols = ['customer_city', 'customer_country', 'customer_zipcode', 'customer_state']

missing_customer_location = df[customer_location_cols].isnull().sum()
unique_customer_location = df[customer_location_cols].nunique()

# Create a DataFrame to display the information
customer_location_info = pd.DataFrame({
    'Missing Values': missing_customer_location,
    'Unique Values': unique_customer_location
})

print("Missing values and unique counts for customer location columns:")
display(customer_location_info)

Missing values and unique counts for customer location columns:


Missing Values  Unique Values
customer_city                793            558
customer_country             757              2
customer_zipcode             750           8866
customer_state               737             44

**Reasoning**:
To impute missing 'customer_city', 'customer_state', and 'customer_country' values using 'customer_zipcode', I will first create mapping dictionaries from known 'customer_zipcode' to their corresponding city, state, and country. Then, I will apply these mappings to fill the missing values.



In [13]:
print("Missing values before zipcode-based imputation:")
print(df[['customer_city', 'customer_state', 'customer_country', 'customer_zipcode']].isnull().sum())

# Create mappings from customer_zipcode to city, state, and country
zipcode_to_city = df.dropna(subset=['customer_zipcode', 'customer_city']).set_index('customer_zipcode')['customer_city'].drop_duplicates().to_dict()
zipcode_to_state = df.dropna(subset=['customer_zipcode', 'customer_state']).set_index('customer_zipcode')['customer_state'].drop_duplicates().to_dict()
zipcode_to_country = df.dropna(subset=['customer_zipcode', 'customer_country']).set_index('customer_zipcode')['customer_country'].drop_duplicates().to_dict()

# Impute missing 'customer_city' using zipcode mapping
df['customer_city'] = df.apply(
    lambda row: zipcode_to_city.get(row['customer_zipcode'], row['customer_city'])
    if pd.isnull(row['customer_city']) and pd.notnull(row['customer_zipcode'])
    else row['customer_city'],
    axis=1
)

# Impute missing 'customer_state' using zipcode mapping
df['customer_state'] = df.apply(
    lambda row: zipcode_to_state.get(row['customer_zipcode'], row['customer_state'])
    if pd.isnull(row['customer_state']) and pd.notnull(row['customer_zipcode'])
    else row['customer_state'],
    axis=1
)

# Impute missing 'customer_country' using zipcode mapping
df['customer_country'] = df.apply(
    lambda row: zipcode_to_country.get(row['customer_zipcode'], row['customer_country'])
    if pd.isnull(row['customer_country']) and pd.notnull(row['customer_zipcode'])
    else row['customer_country'],
    axis=1
)

print("\nMissing values after zipcode-based imputation:")
print(df[['customer_city', 'customer_state', 'customer_country', 'customer_zipcode']].isnull().sum())


Missing values before zipcode-based imputation:
customer_city       793
customer_state      737
customer_country    757
customer_zipcode    750
dtype: int64

Missing values after zipcode-based imputation:
customer_city       492
customer_state      459
customer_country    475
customer_zipcode    750
dtype: int64


**Reasoning**:
To further impute missing 'customer_country' values, I will create a mapping based on known combinations of 'customer_city' and 'customer_state', and then apply this mapping to the DataFrame.



In [14]:
print("Missing values before city/state-based country imputation:")
print(df[['customer_city', 'customer_state', 'customer_country']].isnull().sum())

# Create a mapping from (city, state) to country
city_state_to_country = df.dropna(subset=['customer_city', 'customer_state', 'customer_country']).groupby(['customer_city', 'customer_state'])['customer_country'].agg(lambda x: x.mode()[0] if not x.mode().empty else np.nan).to_dict()

# Impute missing 'customer_country' using city and state mapping
def fill_country_from_city_state(row):
    if pd.isnull(row['customer_country']) and pd.notnull(row['customer_city']) and pd.notnull(row['customer_state']):
        return city_state_to_country.get((row['customer_city'], row['customer_state']), row['customer_country'])
    return row['customer_country']

df['customer_country'] = df.apply(fill_country_from_city_state, axis=1)

print("\nMissing values after city/state-based country imputation:")
print(df[['customer_city', 'customer_state', 'customer_country']].isnull().sum())

Missing values before city/state-based country imputation:
customer_city       492
customer_state      459
customer_country    475
dtype: int64

Missing values after city/state-based country imputation:
customer_city       492
customer_state      459
customer_country     44
dtype: int64


**Reasoning**:
After applying the imputation steps, I need to review the remaining missing values in the customer location columns to assess the effectiveness of the imputation and identify any values that could not be filled. Then, I will summarize the imputation results to complete the subtask.



In [15]:
print("\nRemaining missing values in customer location columns after all imputation steps:")
customer_location_cols = ['customer_city', 'customer_country', 'customer_zipcode', 'customer_state']
missing_after_imputation = df[customer_location_cols].isnull().sum()
missing_percentage_after_imputation = round((df[customer_location_cols].isnull().sum() / len(df)) * 100, 2)

summary_df = pd.DataFrame({
    'Missing Values': missing_after_imputation,
    '% Missing': missing_percentage_after_imputation
})
display(summary_df)

print("\nSummary of imputation results for customer location columns:")
print("Initial missing values for customer_city: 793, customer_state: 737, customer_country: 757, customer_zipcode: 750")
print(f"After zipcode-based imputation, customer_city: {492} missing, customer_state: {459} missing, customer_country: {475} missing, customer_zipcode: {750} missing.")
print(f"After city/state-based country imputation, customer_country: {44} missing.")
print(f"Current missing values - customer_city: {missing_after_imputation['customer_city']}, customer_state: {missing_after_imputation['customer_state']}, customer_country: {missing_after_imputation['customer_country']}, customer_zipcode: {missing_after_imputation['customer_zipcode']}.")


Remaining missing values in customer location columns after all imputation steps:


Missing Values  % Missing
customer_city                492       3.16
customer_country              44       0.28
customer_zipcode             750       4.82
customer_state               459       2.95


Summary of imputation results for customer location columns:
Initial missing values for customer_city: 793, customer_state: 737, customer_country: 757, customer_zipcode: 750
After zipcode-based imputation, customer_city: 492 missing, customer_state: 459 missing, customer_country: 475 missing, customer_zipcode: 750 missing.
After city/state-based country imputation, customer_country: 44 missing.
Current missing values - customer_city: 492, customer_state: 459, customer_country: 44, customer_zipcode: 750.


## Final Task

### Subtask:
Summarize the imputation process for customer location columns and its results.


## Summary:

### Data Analysis Key Findings
*   Initially, `customer_city` had 793 missing values, `customer_country` had 757, `customer_zipcode` had 750, and `customer_state` had 737.
*   After the first imputation step using `customer_zipcode` mappings, missing values significantly reduced: `customer_city` decreased from 793 to 492, `customer_state` from 737 to 459, and `customer_country` from 757 to 475.
*   The second imputation step, which used `customer_city` and `customer_state` combinations to fill missing `customer_country` values, further reduced missing `customer_country` from 475 to 44.
*   Post-imputation, the percentage of missing values for the location columns are: `customer_city` at 3.16\% (492 missing), `customer_country` at 0.28\% (44 missing), `customer_zipcode` at 4.82\% (750 missing), and `customer_state` at 2.95\% (459 missing).

### Insights or Next Steps
*   The multi-stage imputation strategy effectively reduced missing values for `customer_city`, `customer_state`, and particularly `customer_country`, leveraging the hierarchical relationship between location data points.
*   Further investigation could be conducted on the remaining missing values for `customer_city`, `customer_zipcode`, and `customer_state` to determine if other imputation methods or external data sources could be utilized, or if these records should be handled separately due to potential data quality issues.


# Task
## Analyze Order Location Missing Values

### Subtask:
Re-examine the missing values and unique counts for 'order_city', 'order_country', and 'market' to understand the current state before imputation.

**Reasoning**:
Before attempting to impute missing values in 'order_city', 'order_country', and 'market', it is crucial to re-examine their current state. This involves calculating the number of missing values and the count of unique values for each of these columns. This analysis will provide a baseline for assessing the effectiveness of the upcoming imputation steps.

## Analyze Order Location Missing Values

### Subtask:
Re-examine the missing values and unique counts for 'order_city', 'order_country', and 'market' to understand the current state before imputation.


**Reasoning**:
To understand the current state of order location data before imputation, I will select the specified columns, calculate their missing values and unique counts, and then display this information in a structured DataFrame.



In [16]:
order_location_cols = ['order_city','order_state', 'order_country','order_region', 'market']

missing_before = df[order_location_cols].isnull().sum()
missing_before.name = 'Missing_Before'
unique_order_location = df[order_location_cols].nunique()

# Create a DataFrame to display the information
order_location_info = pd.DataFrame({
    'Missing Values': missing_before,
    'Unique Values': unique_order_location
})

print("Missing values and unique counts for order location columns:")
display(order_location_info)

Missing values and unique counts for order location columns:


Missing Values  Unique Values
order_city                736           2717
order_state               770            890
order_country             744            148
order_region              799             23
market                    795              5

In [17]:
# Fonction utilitaire pour créer des dictionnaires "propres" (sans ambiguïté)
def get_clean_mapping(df, key_col, target_col):
    # On prend les paires uniques qui existent
    pairs = df[[key_col, target_col]].dropna().drop_duplicates()

    # On cherche les clés qui pointent vers plusieurs cibles différentes (ex: 'Paris' -> 'TX' et 'Paris' -> 'TN')
    ambiguous_keys = pairs[key_col][pairs[key_col].duplicated()]

    if len(ambiguous_keys) > 0:
        print(f"⚠️ {len(ambiguous_keys)} valeurs ambiguës ignorées pour {key_col} -> {target_col}")

    # On ne garde que les clés sûres (qui n'ont qu'une seule réponse possible)
    clean_pairs = pairs[~pairs[key_col].isin(ambiguous_keys)]
    return clean_pairs.set_index(key_col)[target_col].to_dict()

print("--- DÉBUT IMPUTATION GÉOGRAPHIQUE ---")

# 1. IMPUTATION DE L'ÉTAT (State)
# On utilise la ville, MAIS seulement si la ville n'existe que dans un seul état
city_to_state = get_clean_mapping(df, 'order_city', 'order_state')
missing_before1 = df['order_state'].isnull().sum()
df['order_state'] = df['order_state'].fillna(df['order_city'].map(city_to_state))
print(f"1. États imputés via Villes : {missing_before1 - df['order_state'].isnull().sum()} lignes corrigées")

# 2. IMPUTATION DU PAYS (Country)
# Priorité 1 : L'État définit le pays (ex: California -> USA)
state_to_country = get_clean_mapping(df, 'order_state', 'order_country')
df['order_country'] = df['order_country'].fillna(df['order_state'].map(state_to_country))

# Priorité 2 : La Ville définit le pays (si l'état manquait aussi)
city_to_country = get_clean_mapping(df, 'order_city', 'order_country')
df['order_country'] = df['order_country'].fillna(df['order_city'].map(city_to_country))
print(f"2. Pays imputés : Reste {df['order_country'].isnull().sum()} manquants")

# 3. IMPUTATION DE LA RÉGION (Region)
# Priorité 1 : L'État définit la région (Crucial pour 'West of USA', 'Central Asia' etc.)
state_to_region = get_clean_mapping(df, 'order_state', 'order_region')
df['order_region'] = df['order_region'].fillna(df['order_state'].map(state_to_region))

# Priorité 2 : Le Pays définit la région (Pour les petits pays sans subdivision régionale dans tes données)
country_to_region = get_clean_mapping(df, 'order_country', 'order_region')
df['order_region'] = df['order_region'].fillna(df['order_country'].map(country_to_region))
print(f"3. Régions imputées : Reste {df['order_region'].isnull().sum()} manquants")

# 4. IMPUTATION DU MARCHÉ (Market)
# La région définit le marché (ex: Western Europe -> Europe)
region_to_market = get_clean_mapping(df, 'order_region', 'market')
df['market'] = df['market'].fillna(df['order_region'].map(region_to_market))

# Fallback : Le pays définit le marché (si la région manque encore)
country_to_market = get_clean_mapping(df, 'order_country', 'market')
df['market'] = df['market'].fillna(df['order_country'].map(country_to_market))

print(f"4. Marchés imputés : Reste {df['market'].isnull().sum()} manquants")
print("--- FIN ---")

--- DÉBUT IMPUTATION GÉOGRAPHIQUE ---
1. États imputés via Villes : 695 lignes corrigées
⚠️ 18 valeurs ambiguës ignorées pour order_state -> order_country
2. Pays imputés : Reste 7 manquants
⚠️ 14 valeurs ambiguës ignorées pour order_state -> order_region
⚠️ 3 valeurs ambiguës ignorées pour order_country -> order_region
3. Régions imputées : Reste 2 manquants
4. Marchés imputés : Reste 0 manquants
--- FIN ---


In [18]:
# 2. À exécuter APRÈS l'imputation
missing_after = df[order_location_cols].isnull().sum()
missing_after.name = 'Missing_After'

# 3. Création du Tableau de Performance
report = pd.DataFrame([missing_before, missing_after]).T

# Calculs des métriques
report['Fixed_Rows'] = report['Missing_Before'] - report['Missing_After']
report['Improvement_%'] = (report['Fixed_Rows'] / report['Missing_Before'] * 100).round(2)

# Affichage avec style (Barre de progression visuelle si sur Colab)
print("\n📊 RAPPORT D'IMPUTATION :")
display(report.style.bar(subset=['Improvement_%'], color='#5fba7d', vmin=0, vmax=100))


📊 RAPPORT D'IMPUTATION :


In [19]:
df['shipping_date'] = pd.to_datetime(df['shipping_date'], unit='D', origin='1899-12-30')
df['order_date'] = pd.to_datetime(df['order_date'], unit='D', origin='1899-12-30')

print("Les colonnes 'shipping_date' et 'order_date' ont été converties en format datetime.")
print(df[['shipping_date', 'order_date']].dtypes)

Les colonnes 'shipping_date' et 'order_date' ont été converties en format datetime.
shipping_date    datetime64[ns]
order_date       datetime64[ns]
dtype: object


In [20]:
# Arrondir à la seconde la plus proche (supprime les millisecondes)
df['shipping_date'] = df['shipping_date'].dt.round('s')
df['order_date'] = df['order_date'].dt.round('s')


In [23]:
df[['order_date','shipping_date']].head()

order_date       shipping_date
0 2015-11-09 17:54:14 2015-11-14 17:54:14
1 2015-11-29 23:09:36 2015-12-02 23:09:36
2 2015-02-23 13:12:00 2015-02-26 13:12:00
3 2017-05-26 03:28:48 2017-05-28 03:28:48
4 2017-08-01 14:38:24 2017-08-03 14:38:24

## Shippin mode imputing

In [22]:

# --- 1. PRÉPARATION DES DONNÉES ---
# On travaille sur une copie propre
viz_df = df.dropna(subset=['shipping_mode']).copy()



# Calcul du délai réel en minutes
viz_df['real_delay'] = (viz_df['shipping_date'] - viz_df['order_date']).dt.total_seconds()

# --- 2. GRAPHIQUE 1 : LA DISTRIBUTION (Boxplot) ---
# Permet de voir les plages de jours (Min, Max, Médiane) pour chaque mode
fig_box = px.box(
    viz_df,
    x="shipping_mode",
    y="real_delay",
    color="shipping_mode",
    title="Distribution des délais (Jours) par Mode d'Expédition",
    labels={"real_delay": "Délai Réel (Jours)", "shipping_mode": "Mode d'Expédition"},
    template="plotly_white"
)

fig_box.update_layout(showlegend=False) # Pas besoin de légende ici, l'axe X suffit
fig_box.show()

# --- 3. GRAPHIQUE 2 : LA FRÉQUENCE (Histogramme Groupé) ---
# C'est LE graphique clé pour ta règle d'imputation.
# Il montre s'il y a des chevauchements (ex: est-ce que 'Standard Class' livre parfois en 2 jours ?)
fig_hist = px.histogram(
    viz_df,
    x="real_delay",
    color="shipping_mode",
    barmode="group", # Important : met les barres côte à côte
    title="Nombre de commandes par Délai et par Mode",
    labels={"real_delay": "Délai Réel (Jours)", "count": "Nombre de commandes"},
    template="plotly_white",
    text_auto=True # Affiche les chiffres sur les barres
)

fig_hist.update_layout(xaxis=dict(tickmode='linear')) # Force l'affichage de tous les jours (0, 1, 2, 3...)
fig_hist.show()

# --- 4. TABLEAU DE VÉRITÉ (PANDAS) ---
print("📊 TABLEAU CROISÉ (Mode vs Délai) :")
cross_tab = pd.crosstab(viz_df['real_delay'], viz_df['shipping_mode'])
display(cross_tab)

📊 TABLEAU CROISÉ (Mode vs Délai) :


shipping_mode  First Class  Same Day  Second Class  Standard Class
real_delay                                                        
43200.0                 95        30           207             405
172800.0               800       287           908            2622
259200.0               351       126           426            1445
345600.0               349       110           457            1455
432000.0               349       111           476            1425
518400.0               357        99           409            1462

# Numerical values imputing

In [22]:
# --- 1. CONFIGURATION & ÉTAT INITIAL ---
cols_math = [
    'order_item_product_price', 'product_price',
    'order_item_quantity', 'sales',
    'order_item_discount', 'order_item_discount_rate',
    'order_item_total_amount',
    'order_profit_per_order', 'profit_per_order', 'order_item_profit_ratio'
]

# Capture de l'état initial pour le rapport final
missing_before_financial = df[cols_math].isnull().sum()
missing_before_financial.name = 'Missing_Before'
total_missing_start = missing_before_financial.sum()

print(f"🛑 Total manquants au départ : {total_missing_start}")

# --- 2. DÉFINITION DES FONCTIONS ---

def apply_math_rules_optimized(df):
    """
    Mise à jour avec la relation : Sales = Total / (1 - Rate)
    """

    # --- 1. CALCUL DE SALES (La variable centrale) ---

    # Méthode A : Sales = Price * Qty
    sales_via_price = df['order_item_product_price'] * df['order_item_quantity']

    # Méthode B : Sales = Total + Discount
    sales_via_total = df['order_item_total_amount'] + df['order_item_discount']

    # Méthode C : Sales = Discount / Rate
    sales_via_rate = df['order_item_discount'] / df['order_item_discount_rate'].replace(0, np.nan)

    # --- NOUVELLE MÉTHODE D (Celle que tu as trouvée) ---
    # Sales = Total / (1 - Rate)
    # On évite la division par zéro si le taux est de 100% (1.0)
    denom = (1 - df['order_item_discount_rate']).replace(0, np.nan)
    sales_via_total_rate = df['order_item_total_amount'] / denom

    # APPLICATION : On ajoute .fillna(sales_via_total_rate) à la chaîne
    df['sales'] = df['sales'].fillna(sales_via_price)\
                             .fillna(sales_via_total)\
                             .fillna(sales_via_rate)\
                             .fillna(sales_via_total_rate)  # <--- AJOUT ICI


    # --- 2. CALCUL DU PRIX UNITAIRE ---
    price_via_sales = df['sales'] / df['order_item_quantity'].replace(0, np.nan)
    df['order_item_product_price'] = df['order_item_product_price'].fillna(price_via_sales)

    # --- 3. CALCUL DE LA QUANTITÉ ---
    qty_via_sales = df['sales'] / df['order_item_product_price'].replace(0, np.nan)
    df['order_item_quantity'] = df['order_item_quantity'].fillna(qty_via_sales)

    # --- 4. CALCUL DU DISCOUNT ---
    disc_via_rate = df['order_item_discount_rate'] * df['sales']
    disc_via_total = df['sales'] - df['order_item_total_amount'] # Celle-ci va s'activer dès que Sales est trouvé !
    df['order_item_discount'] = df['order_item_discount'].fillna(disc_via_rate).fillna(disc_via_total)

    # --- 5. CALCUL DU DISCOUNT RATE ---
    rate_via_disc = df['order_item_discount'] / df['sales'].replace(0, np.nan)
    # Nouvelle astuce : Rate = 1 - (Total / Sales)
    rate_via_total = 1 - (df['order_item_total_amount'] / df['sales'].replace(0, np.nan))

    df['order_item_discount_rate'] = df['order_item_discount_rate'].fillna(rate_via_disc).fillna(rate_via_total)

    # --- 6. CALCUL DU TOTAL AMOUNT ---
    total_via_sales = df['sales'] - df['order_item_discount']
    # Nouvelle astuce inverse : Total = Sales * (1 - Rate)
    total_via_rate = df['sales'] * (1 - df['order_item_discount_rate'])
    total_via_profit = df['order_profit_per_order'] / df['order_item_profit_ratio'].replace(0, np.nan)

    df['order_item_total_amount'] = df['order_item_total_amount'].fillna(total_via_sales)\
                                                                 .fillna(total_via_rate)\
                                                                 .fillna(total_via_profit)

    # --- 7. CALCUL DU PROFIT ---
    profit_via_ratio = df['order_item_profit_ratio'] * df['order_item_total_amount']
    df['order_profit_per_order'] = df['order_profit_per_order'].fillna(profit_via_ratio)

    # --- 8. CALCUL DU PROFIT RATIO ---
    ratio_via_profit = df['order_profit_per_order'] / df['order_item_total_amount'].replace(0, np.nan)
    df['order_item_profit_ratio'] = df['order_item_profit_ratio'].fillna(ratio_via_profit)

    return df

def run_math_loop(df, phase_name):
    """Exécute la boucle mathématique jusqu'à convergence."""
    print(f"\n🚀 --- DÉBUT {phase_name} ---")
    max_iterations = 10
    for i in range(max_iterations):
        missing_iter_start = df[cols_math].isnull().sum().sum()

        df = apply_math_rules_optimized(df)

        missing_iter_end = df[cols_math].isnull().sum().sum()
        print(f"   🔄 Itération {i+1}: Manquants {missing_iter_start} -> {missing_iter_end}")

        if missing_iter_start == missing_iter_end:
            print("   ✅ Convergence atteinte.")
            break
    return df

# --- 3. EXÉCUTION DE LA STRATÉGIE "SANDWICH" ---

# ÉTAPE A : Fusion des Redondances (Pré-requis)
print("\n[1/4] Fusion des colonnes redondantes...")
df['order_item_product_price'] = df['order_item_product_price'].fillna(df['product_price'])
df['order_profit_per_order'] = df['order_profit_per_order'].fillna(df['profit_per_order'])

# ÉTAPE B : Phase 1 - Maths Préliminaires
# On résout tout ce qui est logique AVANT de faire des estimations
df = run_math_loop(df, "PHASE 1 : MATHS PURES")

# ÉTAPE C : Phase 2 - Imputation Statistique (Le Prix par Produit)
print("\n[3/4] Imputation Statistique (Prix Médian par ID Produit)...")
missing_price_before = df['order_item_product_price'].isnull().sum()

# On utilise les prix trouvés en Phase 1 pour renforcer le calcul du médian
price_map = df.groupby('product_card_id')['order_item_product_price'].transform('median')
df['order_item_product_price'] = df['order_item_product_price'].fillna(price_map)

missing_price_after = df['order_item_product_price'].isnull().sum()
print(f"   📉 Prix récupérés via estimation : {missing_price_before - missing_price_after}")

# ÉTAPE D : Phase 3 - Maths Finales
# On relance les maths pour déduire Sales/Totals grâce aux nouveaux prix estimés
df = run_math_loop(df, "PHASE 3 : MATHS FINALES")

# --- 4. RAPPORT FINAL ---
missing_after_financial = df[cols_math].isnull().sum()
missing_after_financial.name = 'Missing_After'

report = pd.DataFrame([missing_before_financial, missing_after_financial]).T
report['Fixed'] = report['Missing_Before'] - report['Missing_After']
report['Improvement_%'] = (report['Fixed'] / report['Missing_Before'] * 100).round(2)

# FILTRAGE : Exclusion des colonnes redondantes pour l'affichage uniquement
rows_to_exclude = ['product_price', 'profit_per_order']
report_clean = report.drop(index=rows_to_exclude, errors='ignore')

print("\n📊 TABLEAU DE SCORE FINAL :")
display(report_clean.style.bar(subset=['Improvement_%'], color='#5fba7d', vmin=0, vmax=100))

🛑 Total manquants au départ : 7803

[1/4] Fusion des colonnes redondantes...

🚀 --- DÉBUT PHASE 1 : MATHS PURES ---
   🔄 Itération 1: Manquants 6303 -> 1524
   🔄 Itération 2: Manquants 1524 -> 1520
   🔄 Itération 3: Manquants 1520 -> 1520
   ✅ Convergence atteinte.

[3/4] Imputation Statistique (Prix Médian par ID Produit)...
   📉 Prix récupérés via estimation : 5

🚀 --- DÉBUT PHASE 3 : MATHS FINALES ---
   🔄 Itération 1: Manquants 1515 -> 1510
   🔄 Itération 2: Manquants 1510 -> 1510
   ✅ Convergence atteinte.

📊 TABLEAU DE SCORE FINAL :


In [23]:
# Liste des colonnes où il reste des trous (j'ai exclu celle à 0)
cols_with_nan = [
    'order_item_product_price',
    'order_item_quantity',
    'sales',
    'order_item_discount',
    'order_item_discount_rate',
    'order_item_total_amount',
    'order_profit_per_order',
    'order_item_profit_ratio'
]

# On filtre : "Montre-moi les lignes où au moins une de ces colonnes est vide"
missing_rows = df[df[cols_with_nan].isnull().any(axis=1)]

print(f"Nombre de lignes trouvées : {len(missing_rows)}")

# Affichage avec mise en évidence des valeurs manquantes (si tu es sur Colab/Jupyter)
# Les cases vides apparaîtront en rouge ou surlignées
display(missing_rows[cols_with_nan].style.highlight_null(color='pink'))

Nombre de lignes trouvées : 2


Columns_imputed = ["category_id